### PR TITLE
Allow ingesters of gen_backend_stubs codegen to customize "class name"

### PR DESCRIPTION
**Pull request by guoyejun**
When a backend (and its corresponding dispatch key) like `XPU` or `Lazy` runs the external codegen pipeline, all of its kernels get generated as methods inside of a `{DispatchKey}NativeFunctions` class. This PR allows the backend to customize the generated name if they wish to.

### CODE REVIEW

**Comment by ngimel** [comment]
Please provide a description of what this PR does, and what's the motivation

**Comment by guoyejun** [comment]
sure, 

this PR is to add an option class_name for external backend yaml file

the motivation is that an existing external backend might already use another class name, and the external backend is not feasible to change the class name due to too many changes, and so we need to support customized class name for generated code.

and the above was also described in the commit log.

**Comment by guoyejun** [comment]
any comment? thanks.

File: tools/codegen/gen_backend_stubs.py [hunk 1]
```diff
@@ -300,17 +303,20 @@ def make_file_manager(install_dir: str) -> FileManager:
         # This could be useful if a backend wants to quickly set up a noop yaml file but doesn't have any kernels ready yet.
         return
 
-    class_name = backend_indices[backend_key].native_function_class_name()
+    if class_name is None:
```
Comment (bdhirsh) [inline tools/codegen/gen_backend_stubs.py:306]: Can you add a detailed comment here similar to what you described in the PR? Just to make it clear that the `class_name` yaml argument is completely optional, but if specified it allows you to override the name of the class that all generated kernel definitions live under.

This bit of the code is probably the closest we have to documentation for the yaml, and we probably want to make it clear that the argument is totally optional (versus some other arguments that toggle important behavior).

Comment (guoyejun) [reply tools/codegen/gen_backend_stubs.py:308]: sure, done, thanks.

**Review by bdhirsh — APPROVED** [review]

File: tools/codegen/gen_backend_stubs.py [hunk 2]
```diff
@@ -133,13 +135,14 @@ def create_backend_index(
 autograd key. They cannot be mix and matched. If this is something you need, feel free to create an issue! \
 {forward_kernels[0].kernel} is listed under "supported", but {backward_kernels[0].kernel} is listed under "autograd".'
 
-    return ParsedExternalYaml(backend_key, autograd_key, cpp_namespace, backend_indices)
+    return ParsedExternalYaml(backend_key, autograd_key, class_name, cpp_namespace, backend_indices)
```
Comment (bdhirsh) [inline tools/codegen/gen_backend_stubs.py:138]: Oh @guoyejun this breaks flake8 lint (see the CI failures in the PR)

Comment (guoyejun) [reply tools/codegen/gen_backend_stubs.py:138]: sure, will update, thanks.

**Comment by bdhirsh** [comment]
Fixed. But as @malfet pointed out to me - it's too late to fix up the corresponding commit message in master :( sorry about that.

**Comment by guoyejun** [comment]
My wrong assumption that the two commits in this PR will be merged with two commits.  Each commit has a meaningful commit title and commit description, but looks that they are merged with the PR title and PR description left.

I will next create one PR for one commit, thanks.

**Pull request closed without merging**